### ISSUE TITLE
Empty hypotheses in HLG decoding.

### ISSUE THREAD

**Issue by videodanchik** [k2-fsa/icefall#292]
Hi, I have a problem with HLG decoding on my test set. Some of the utterances are producing an empty `hyp`. I'm giving a link to download my `HLG` graph, which was used to generate a lattice for a particular utterance, that produces an empty `hyp`, along with the `nnet_output` tensor. Can you please help to figure out the underlying reason for this bug? 

`nnet_output`
https://drive.google.com/file/d/1kakVVnhGIYtSvi3bGoaaFA-4Yg-Nx_o5/view?usp=sharing

`HLG`
https://drive.google.com/file/d/1CqiuSTVdQjGxYWAunrqS-YU9fClBArtG/view?usp=sharing

**Comment by pkufool** [comment]
Sorry, I can't load your lattice by `torch.load("lattice.pt")`,  which says `RecursionError: maximum recursion depth exceeded`. Can you confirm that your lattice ia valid.

> Also, how can I visualize the lattice properly in this case. I tried to execute lattice[0].draw('lattice.png'), but wasn't able to see anything as there seems to be too many nodes in the resulting graph.

You can try drawing the lattice as pdf  by `lattice[0].draw('lattice.pdf')`.

**Comment by videodanchik** [comment]
@pkufool Thanks for response! I'm experiencing the same error as you, sorry I didn't check if I can read it after saving. Is there any other way for me to share the lattice with you? Does `torch.save(lattice, 'lattice.pt')` works for you, because I can't save any lattice with that?

> You can try drawing the lattice as pdf by lattice[0].draw('lattice.pdf').
Thanks that has worked! I'm attaching the `lattice.pdf` that I have generated.
[lattice.pdf](https://github.com/k2-fsa/icefall/files/8447458/lattice.pdf)

**Comment by pkufool** [comment]
>  Does torch.save(lattice, 'lattice.pt') works for you, because I can't save any lattice with that?

You should convert the fsa to a dict first.
```
torch.save(lattice.as_dict(), 'lattice.pt')
```
Then load it by:
```
lattice = k2.Fsa.from_dict(torch.load("lattice.pt"))
```

BTW, There are too many nodes in lattice.pdf, try using connect  to remove the un-accessible states.
You can do as follows:
```
lattice = k2.connect(lattice)
lattice[0].draw('lattice.pdf')
```

**Comment by videodanchik** [comment]
Cool thanks! So `k2.connect(lattice).as_dict()` returns a trivial lattice

```
{'arcs': tensor([1, 0, 0, 0, 0, 0], dtype=torch.int32),
 'aux_labels': RaggedTensor([], dtype=torch.int32),
 'tokens': tensor([], dtype=torch.int32),
 'lm_scores': tensor([])}
```
I've saved the original lattice as you said and attached it
[lattice.pt.zip](https://github.com/k2-fsa/icefall/files/8447948/lattice.pt.zip)

So it seems that when `search_beam` is not enough and it can produce an unconnected lattice. I was decoding with the following params to generate it

```
lattice = get_lattice(
    nnet_output=nnet_output,
    decoding_graph=HLG,
    supervision_segments=supervision_segments,
    search_beam=12.0,
    output_beam=4.0,
    min_active_states=100,
    max_active_states=7000,
    subsampling_factor=4,
)

```

simply increasing `search_beam` to 15 solves the issue and `best_path` gives a good transcript. Can give some comments, why is it happening? How would you recommend to deal with that? I can iterate over a range of `search_beams` until it outputs a transcript, but it doesn't sound like the best approach. I'm curious how Kaldi chain models HLG decoding resolve this issue or it doesn't occur?

**Comment by danpovey** [comment]
I am working on some instructions on how to implement "force out" or --allow-partial decoding.

**Comment by danpovey** [comment]
In Kaldi there is an option --allow-partial which allows you to accept any path if it has a route to the final state.
It looks to me like this may be complex to implement in k2.
What I think might be a reasonable substitute, is to change the pruning behavior on the last frame (the frame
that has -1's) and possibly a few frames before that, e.g. the 5 last frames.  The change to 
lambda_set_beam_and_cutoffs would be something like:
```
  int32_t final_t = b_fsas_row_splits1[i+1] - b_fsas_row_splits1[i];
  if (t + 5 >= final_t) {  
       // Do less pruning on the final few frames, to ensure we don't prune away final states.
       min_active = max(min_active, max_active/2);
  }
  // ... then all the existing logic..
  if (t == final_t)  // no pruning on last frame; we want all final-arcs.     
      dynamic_beam = 1.0e+10;
```

**Comment by danpovey** [comment]
... this will not fully guarantee that there are active states on the last frame, it will just make it extremely rare, so we should make sure to not crash if the lattice is empty after connect().

**Comment by videodanchik** [comment]
Hi @danpovey, sorry, could you provide more details on where exactly these lines should be inserted (line numbers), because I tried to paste this code into different places and got compilation errors. I'm not an experienced c++ developer, so please make a discount to that).

**Comment by danpovey** [comment]
It would be great if you could do a PR; the instructions were actually intended for someone like @pkufool but it's great that you were able to figure it out.

**Comment by pkufool** [comment]
Sorry, I was busy with other stuff, did not pay much attention here. @videodanchik  you can do it if you have time.

**Comment by pkufool** [comment]
@videodanchik  Do you have time to fix this issue, if not, I will file a PR. Thanks!

**Comment by videodanchik** [comment]
@pkufool Sorry, after additional decoding of my test utterances I understood, that I've made a mistake somewhere in the code, because the number of empty utterances doubled. Now with your PR I see that
```
if (t + 5 < final_t) {
    current_min_active = min(min_active, max_active / 2);
}
```
while Dan's original suggestion was `min_active = max(min_active, max_active/2);` so I missed `max` with `min` and after increasing `search_beam` to 15.0 I just started getting very few of these empty utterances (without any code modifications), like < 0.5% for any of my test dataset, so that was ok for me, I just moved on and didn't close the issue or at least I should have asked what did I do wrong in my C++ modifications. Thanks for fixing this, I'll try it out, it should decrease WER slightly. BTW, did you try to decode any of the test sets with your modification?
